libkbfs/prev_revisions.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"github.com/keybase/go-codec/codec"
+	"github.com/keybase/kbfs/kbfsmd"
+)
+
+// PrevRevisionAndCount track the MD version of a previous revision of
+// a dir entry, and how many revisions ago that was from the current
+// revision.
+type PrevRevisionAndCount struct {
+	Revision kbfsmd.Revision `codec:"r"`
+	Count    uint8           `codec:"c"`
+
+	codec.UnknownFieldSetHandler
+}
+
+// minPrevRevisionSlotCounts defines the "min count" of each
+// corresponding entry in a `PrevRevisions` slice.  The length of
+// `minPrevRevisionSlotCounts` is the max length of a `PrevRevisions`
+// slice.
+var minPrevRevisionSlotCounts = [...]uint8{1, 5, 20, 50, 100}
+
+// PrevRevisions tracks several previous versions of a file in order
+// of descending revision number, starting with the most recent.
+type PrevRevisions []PrevRevisionAndCount
+
+// addRevision returns a copy of `pr` with a new immediately-previous
+// revision added, with the existing entries moved or overwritten to
+// accomodate the new entry, and with increased counts.  Any existing
+// revisions smaller than or equal to minRev will be removed.
+func (pr PrevRevisions) addRevision(
+	r, minRev kbfsmd.Revision) (ret PrevRevisions) {
+	newLength := len(pr)
+	if newLength < len(minPrevRevisionSlotCounts) {
+		newLength++
+	}
+	ret = make(PrevRevisions, newLength)
+	copy(ret, pr)
+	earliestGoodSlot := 0
+	numDropped := 0
+
+	for i, prc := range ret {
+		if prc.Count == 255 {
+			panic("Previous revision count is about to overflow")
```

Comment:
It seems like this is a distinct possibility. Shouldn't we do something other than panic?

Reply:
I'll add a comment, but I don't think it's possible for the count to go to 255.  Outline of proof:

* The count in a slot only gets incremented when there's a new revision of the file.
* As soon as a slot has the count equal to the min-count of the slot to its right, it gets moved over, replacing whatever was there before.
* So the question is, once something gets to 100, how many file-revisions will it be before something else overwrites it?
* Someone better at math than me can probably figure out an exact answer, but intuitively the answer is less than 150, given the other min slot counts of 1,5,20,50.

Also, I modified the playground code from before to panic if it reaches 255, and set it to run for 100,000 file revisions, and it doesn't panic: https://play.golang.org/p/ltVzr1PH-MG.  The max count it sees is 178.  (It might be a partial coincidence, but that's the sum of all the slots, + 2.